core/src/findlib/hardlink.h

```diff
@@ -21,18 +21,68 @@
 #ifndef BAREOS_FINDLIB_HARDLINK_H_
 #define BAREOS_FINDLIB_HARDLINK_H_
 
-CurLink* lookup_hardlink(JobControlRecord* jcr,
-                         FindFilesPacket* ff_pkt,
-                         ino_t ino,
-                         dev_t dev);
-CurLink* new_hardlink(JobControlRecord* jcr,
-                      FindFilesPacket* ff_pkt,
-                      char* fname,
-                      ino_t ino,
-                      dev_t dev);
-void FfPktSetLinkDigest(FindFilesPacket* ff_pkt,
-                        int32_t digest_stream,
-                        const char* digest,
-                        uint32_t len);
+#include <unordered_map>
+#include <vector>
+#include <string>
+#include <type_traits>
+
+/**
+ * Structure for keeping track of hard linked files, we
+ * keep an entry for each hardlinked file that we save,
+ * which is the first one found. For all the other files that
+ * are linked to this one, we save only the directory
+ * entry so we can link it.
+ */
+struct CurLink {
+  uint32_t FileIndex{0};      /**< Bareos FileIndex of this file */
+  int32_t digest_stream{0};   /**< Digest type if needed */
+  std::vector<char> digest{}; /**< Checksum of the file if needed */
+  std::string name;           /**< The name */
+
+  CurLink(const char* fname) : name{fname} {}
+  void set_digest(int32_t new_digest_stream,
+                  const char* new_digest,
+                  uint32_t len)
+  {
+    if (digest.empty()) {
+      digest.assign(new_digest, new_digest + len);
+      digest_stream = new_digest_stream;
+    }
+  }
+};
+
+// workaround for the win32 compat layer defining struct stat
+// without dev_t/ino_t
+using our_dev_t = decltype(static_cast<struct stat*>(nullptr)->st_dev);
+using our_ino_t = decltype(static_cast<struct stat*>(nullptr)->st_ino);
+
+// sometimes decltype is weird.  Make sure
+// that we ended up with T and not T&
+static_assert(!std::is_reference_v<our_dev_t>);
+static_assert(!std::is_reference_v<our_ino_t>);
+
+struct Hardlink {
+  our_dev_t dev; /**< Device */
+  our_ino_t ino; /**< Inode with device is unique */
+
+  friend bool operator==(const Hardlink& l, const Hardlink& r)
+  {
+    return l.dev == r.dev && l.ino == r.ino;
+  }
+};
+
+template <> struct std::hash<Hardlink> {
+  std::size_t operator()(const Hardlink& link) const
+  {
+    auto hash1 = std::hash<decltype(link.dev)>{}(link.dev);
+    auto hash2 = std::hash<decltype(link.ino)>{}(link.ino);
+
+    // change this when N3876 (std::hash_combine) or something similar
+    // is finally implemented.
+    return hash1 + 67 * hash2;
```

Comment:
I think we should either import the 3-4 functions from N3876 page 2/3 or at least use the proposed algorithm from page 2 here (and keep a reference to the paper).

```suggestion
    const auto hash1 = std::hash<decltype(link.dev)>{}(link.dev);
    const auto hash2 = std::hash<decltype(link.ino)>{}(link.ino);

    // algorithm proposed in N3876 (std::hash_combine)
    size_t seed{0};
    seed ^= hash1 + 0x9e3779b9 + (seed<<6) + (seed>>2);    
    seed ^= hash2 + 0x9e3779b9 + (seed<<6) + (seed>>2);
    return seed;
```

Reply:
I am not sure with what license the code in the paper was released, so to stay on the safe side i used a different combinator.

Reply:
I added the function.  Since the function originally comes from boost, should we add the boost license to the license file ?

Reply:
As discussed, I don't think there is enough level of creativity in there to require us to attribute boost. Also, the boost people seem to be unsure where this originated from.

If you want to address the different `sizeof(size_t)` cases, you can probably just `if constexpr` that, but I doubt it will help a lot.